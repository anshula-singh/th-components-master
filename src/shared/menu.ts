import { produce } from "immer";
import { ShadowRootTheGoodPart } from "lwc";

type Input = {
  onChange: (state: MenuState) => void;
  shadowRoot: ShadowRootTheGoodPart;
};

export type MenuState = {
  count: number;
  open: boolean;
  selectedIndex: number;
  setup: boolean;
};

export type Selector = "trigger" | "menu" | "item";

const SELECTOR_TRIGGER = "trigger";
const SELECTOR_MENU = "menu";
const SELECTOR_ITEM = "item";

export function getSelector(selector: Selector) {
  return `[data-menu="${selector}"]`;
}

export class Menu {
  private _state: MenuState = {
    count: 0,
    open: false,
    selectedIndex: -1,
    setup: false
  };
  private get state(): MenuState {
    return this._state;
  }
  private set state(value: MenuState) {
    this._state = value;
    this.update();
    this.input.onChange(this._state);
  }

  constructor(private input: Input) {
    input.onChange(this.state);
  }

  private setState(
    nextState: Partial<MenuState> | ((state: MenuState) => void)
  ) {
    this.state = produce(this.state, state => {
      if (typeof nextState === "function") nextState(state);
      else Object.assign(state, nextState);
    });
  }

  renderedCallback() {
    if (!this.state.setup) {
      let $trigger = this.getElement(SELECTOR_TRIGGER)!;
      let $menu = this.getElement(SELECTOR_MENU)!;
      $trigger.addEventListener("click", this.onTriggerClick);
      $trigger.addEventListener("keydown", this.onTriggerKeyDown);
      $menu.addEventListener("blur", this.onMenuBlur);
      $menu.addEventListener("click", this.onMenuClick);
      $menu.addEventListener("keydown", this.onMenuKeyDown);
      this.state.setup = true;
    }
    this.state.count = this.input.shadowRoot.querySelectorAll(
      getSelector(SELECTOR_ITEM)
    ).length;
    this.update();
  }

  disconnectedCallback() {
    let $trigger = this.getElement(SELECTOR_TRIGGER)!;
    let $menu = this.getElement(SELECTOR_MENU)!;
    $trigger.removeEventListener("click", this.onTriggerClick);
    $trigger.removeEventListener("keydown", this.onTriggerKeyDown);
    $menu.removeEventListener("blur", this.onMenuBlur);
    $menu.removeEventListener("click", this.onMenuClick);
    $menu.removeEventListener("keydown", this.onMenuKeyDown);
  }

  open() {
    this.setState({
      open: true,
      selectedIndex: 0
    });
    this.getElement(SELECTOR_MENU)!.focus();
  }

  close() {
    this.setState({ open: false });
  }

  private getElement(selector: Selector) {
    return this.input.shadowRoot.querySelector(getSelector(selector));
  }

  private onTriggerClick = (e: MouseEvent) => {
    this.open();
  };

  private onTriggerKeyDown = (e: KeyboardEvent) => {
    switch (e.key) {
      case "Enter":
      case "Space":
      case "ArrowDown":
        this.open();
        break;
    }
  };

  private onMenuClick = (e: MouseEvent) => {};

  private onMenuBlur = (e: FocusEvent) => {
    setTimeout(() => {
      let $active = this.input.shadowRoot.activeElement;
      let $menu = this.getElement(SELECTOR_MENU)!;
      if ($menu === $active || $menu.contains($active)) return;
      this.setState({ open: false });
    });
  };

  private onMenuKeyDown = (e: KeyboardEvent) => {
    switch (e.key) {
      case "ArrowUp":
        this.setState(state => {
          state.selectedIndex--;
          if (state.selectedIndex < 0) state.selectedIndex = state.count - 1;
        });
        break;
      case "ArrowDown":
        this.setState(state => {
          state.selectedIndex++;
          if (state.selectedIndex === state.count) state.selectedIndex = 0;
        });
        break;
      case "Enter":
        if (!this.state.open) break;
        this.close();
        this.select();
        break;
      case "Escape":
      case "Esc": {
        this.close();
        this.getElement(SELECTOR_TRIGGER)!.focus();
        break;
      }
    }
  };

  private select() {
    let element = this.input.shadowRoot.querySelectorAll(
      getSelector(SELECTOR_ITEM)
    )[this.state.selectedIndex] as HTMLElement | undefined;
    if (element) element.click();
  }

  private setTriggerAttributes() {
    let $trigger = this.getElement(SELECTOR_TRIGGER)!;
    let $menu = this.getElement(SELECTOR_MENU)!;
    $trigger.setAttribute("aria-controls", $menu.getAttribute("id") || "");
    $trigger.setAttribute("aria-expanded", String(this.state.open));
    $trigger.setAttribute("aria-haspopup", "true");
    $trigger.setAttribute("role", "button");
  }

  private setMenuAttributes() {
    let $menu = this.getElement(SELECTOR_MENU)!;
    let $activeDecendant = this.input.shadowRoot.querySelectorAll(
      getSelector(SELECTOR_ITEM)
    )[this.state.selectedIndex] as HTMLElement | undefined;
    $menu.setAttribute(
      "aria-activedescendant",
      $activeDecendant ? $activeDecendant.getAttribute("id") || "" : ""
    );
    $menu.setAttribute("aria-hidden", String(!this.state.open));
    $menu.style.zIndex = this.state.open ? "100" : "";
    $menu.style.display = this.state.open ? "block" : "none";
    if (this.state.open) {
      $menu.setAttribute("tabindex", "0");
    } else {
      $menu.removeAttribute("tabindex");
    }
    this.input.shadowRoot.querySelectorAll("ui,li").forEach(node => {
      let element = node as HTMLElement;
      element.setAttribute("role", "presentation");
    });
    this.input.shadowRoot
      .querySelectorAll(getSelector(SELECTOR_ITEM))
      .forEach(node => {
        let element = node as HTMLElement;
        element.setAttribute("tabindex", "-1");
      });
  }

  private setMenuItemsAttributes() {
    this.input.shadowRoot
      .querySelectorAll(getSelector(SELECTOR_ITEM))
      .forEach((node, i) => {
        let element = node as HTMLElement;
        element.setAttribute(
          "data-active",
          String(i === this.state.selectedIndex)
        );
        element.setAttribute("role", "menuitem");
      });
  }

  private update() {
    this.setTriggerAttributes();
    this.setMenuAttributes();
    this.setMenuItemsAttributes();
  }
}
