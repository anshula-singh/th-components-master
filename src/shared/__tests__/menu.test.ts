import { Menu } from "../menu";

jest.useFakeTimers();

describe("Menu", () => {
  let $shadow: ShadowRoot;
  let $trigger: HTMLElement;
  let $menu: HTMLElement;
  let menu: Menu;
  afterEach(() => {
    while (document.body.firstChild) {
      document.body.removeChild(document.body.firstChild);
    }
  });
  beforeEach(() => {
    let $element = document.createElement("div");
    $shadow = $element.attachShadow({ mode: "open" });
    $shadow.innerHTML = createMenuTemplate(createMenuItems());
    menu = createMenu($shadow);
    menu.renderedCallback();
    $trigger = $shadow.querySelector(`[data-menu="trigger"]`) as HTMLElement;
    $menu = $shadow.querySelector(`[data-menu="menu"]`) as HTMLElement;
  });
  it("default", async () => {
    expect($shadow.innerHTML).toMatchSnapshot();
  });
  describe("trigger", () => {
    it("click", async () => {
      let event = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      $trigger.dispatchEvent(event);
      expect($shadow.innerHTML).toMatchSnapshot();
    });
    it("enter", async () => {
      let event = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "Enter"
      });
      $trigger.dispatchEvent(event);
      expect($shadow.innerHTML).toMatchSnapshot();
    });
    it("space", async () => {
      let event = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "Space"
      });
      $trigger.dispatchEvent(event);
      expect($trigger).toHaveAttribute("aria-expanded", "true");
    });
    it("arrow down", async () => {
      let event = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "ArrowDown"
      });
      $trigger.dispatchEvent(event);
      expect($trigger).toHaveAttribute("aria-expanded", "true");
    });
  });
  describe("menu", () => {
    it("blur (no active child)", () => {
      let clickEvent = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      let blurEvent = new FocusEvent("blur", {
        bubbles: false,
        cancelable: true
      });
      expect($trigger).toHaveAttribute("aria-expanded", "false");
      $trigger.dispatchEvent(clickEvent);
      expect($trigger).toHaveAttribute("aria-expanded", "true");
      $menu.dispatchEvent(blurEvent);
      Object.defineProperty($shadow, "activeElement", { get: () => null });
      jest.runAllTimers();
      expect($trigger).toHaveAttribute("aria-expanded", "false");
    });
    it("blur (active chilld)", () => {
      let clickEvent = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      let blurEvent = new FocusEvent("blur", {
        bubbles: false,
        cancelable: true
      });
      expect($trigger).toHaveAttribute("aria-expanded", "false");
      $trigger.dispatchEvent(clickEvent);
      expect($trigger).toHaveAttribute("aria-expanded", "true");
      $menu.dispatchEvent(blurEvent);
      Object.defineProperty($shadow, "activeElement", {
        get: () => $shadow.querySelector("li")
      });
      jest.runAllTimers();
      expect($trigger).toHaveAttribute("aria-expanded", "true");
    });
    it("escape", () => {
      let clickEvent = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      let keydownEvent = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "Escape"
      });
      expect($trigger).toHaveAttribute("aria-expanded", "false");
      $trigger.dispatchEvent(clickEvent);
      expect($trigger).toHaveAttribute("aria-expanded", "true");
      $menu.dispatchEvent(keydownEvent);
      expect($trigger).toHaveAttribute("aria-expanded", "false");
    });
    it("arrow down", () => {
      let clickEvent = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      let keydownEvent = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "ArrowDown"
      });
      $trigger.dispatchEvent(clickEvent);
      $menu.dispatchEvent(keydownEvent);
      expect($menu).toHaveAttribute("aria-activedescendant", "b");
      $menu.dispatchEvent(keydownEvent);
      expect($menu).toHaveAttribute("aria-activedescendant", "c");
    });
    it("arrow up", () => {
      let clickEvent = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      let keydownEvent = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "ArrowUp"
      });
      $trigger.dispatchEvent(clickEvent);
      $menu.dispatchEvent(keydownEvent);
      expect($menu).toHaveAttribute("aria-activedescendant", "c");
      $menu.dispatchEvent(keydownEvent);
      expect($menu).toHaveAttribute("aria-activedescendant", "b");
    });
    it("enter", () => {
      let clickEvent = new MouseEvent("click", {
        bubbles: false,
        cancelable: true
      });
      let keydownEvent = new KeyboardEvent("keydown", {
        bubbles: false,
        cancelable: true,
        key: "Enter"
      });
      let onClick = jest.fn();
      $menu.querySelector("a")!.addEventListener("click", onClick);
      $trigger.dispatchEvent(clickEvent);
      $menu.dispatchEvent(keydownEvent);
      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
});

const createMenuTemplate = (
  items: { href: string; key: string; label: string }[]
) => `
  <button data-menu="trigger">
    Profile
  </button>
  <div data-menu="menu" id="menu">
    <ul>
      ${items.map(
        item => `
        <li>
          <a data-menu="item" href="${item.href}" id="${item.key}">
            ${item.label}
          </a>
        </li>
      `
      )}
    </ul>
  </div>
`;

const createMenuItems = () => [
  { href: "/a", key: "a", label: "A" },
  { href: "/b", key: "b", label: "B" },
  { href: "/c", key: "c", label: "C" }
];

const createMenu = (shadowRoot: ShadowRoot) =>
  new Menu({
    onChange: state => {},
    // @ts-ignore
    shadowRoot: shadowRoot
  });
