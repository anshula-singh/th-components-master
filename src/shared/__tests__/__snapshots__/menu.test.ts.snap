// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu default 1`] = `
"
  <button data-menu=\\"trigger\\" aria-controls=\\"menu\\" aria-expanded=\\"false\\" aria-haspopup=\\"true\\" role=\\"button\\">
    Profile
  </button>
  <div data-menu=\\"menu\\" id=\\"menu\\" aria-activedescendant=\\"\\" aria-hidden=\\"true\\" style=\\"display: none;\\">
    <ul>
      
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/a\\" id=\\"a\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            A
          </a>
        </li>
      ,
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/b\\" id=\\"b\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            B
          </a>
        </li>
      ,
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/c\\" id=\\"c\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            C
          </a>
        </li>
      
    </ul>
  </div>
"
`;

exports[`Menu trigger click 1`] = `
"
  <button data-menu=\\"trigger\\" aria-controls=\\"menu\\" aria-expanded=\\"true\\" aria-haspopup=\\"true\\" role=\\"button\\">
    Profile
  </button>
  <div data-menu=\\"menu\\" id=\\"menu\\" aria-activedescendant=\\"a\\" aria-hidden=\\"false\\" style=\\"display: block; z-index: 100;\\" tabindex=\\"0\\">
    <ul>
      
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/a\\" id=\\"a\\" tabindex=\\"-1\\" data-active=\\"true\\" role=\\"menuitem\\">
            A
          </a>
        </li>
      ,
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/b\\" id=\\"b\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            B
          </a>
        </li>
      ,
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/c\\" id=\\"c\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            C
          </a>
        </li>
      
    </ul>
  </div>
"
`;

exports[`Menu trigger enter 1`] = `
"
  <button data-menu=\\"trigger\\" aria-controls=\\"menu\\" aria-expanded=\\"true\\" aria-haspopup=\\"true\\" role=\\"button\\">
    Profile
  </button>
  <div data-menu=\\"menu\\" id=\\"menu\\" aria-activedescendant=\\"a\\" aria-hidden=\\"false\\" style=\\"display: block; z-index: 100;\\" tabindex=\\"0\\">
    <ul>
      
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/a\\" id=\\"a\\" tabindex=\\"-1\\" data-active=\\"true\\" role=\\"menuitem\\">
            A
          </a>
        </li>
      ,
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/b\\" id=\\"b\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            B
          </a>
        </li>
      ,
        <li role=\\"presentation\\">
          <a data-menu=\\"item\\" href=\\"/c\\" id=\\"c\\" tabindex=\\"-1\\" data-active=\\"false\\" role=\\"menuitem\\">
            C
          </a>
        </li>
      
    </ul>
  </div>
"
`;
